src/portaudiosound.cpp

```diff
@@ -0,0 +1,279 @@
+    CSoundBase ( "portaudio", fpNewProcessCallback, arg, strMIDISetup ),
+    deviceIndex (-1),
+    deviceStream (NULL)
+{
+    pThisSound = this;
+
+    PaError err = Pa_Initialize ();
+    if ( err != paNoError )
+    {
+        throw CGenErr( tr( "Failed to initialize PortAudio: %1" ).arg ( Pa_GetErrorText ( err )) );
+    }
+
+    PaAsio_RegisterBufferSizeChangeCallback ( &bufferSizeChangeCallback );
+    PaAsio_RegisterResetRequestCallback ( &resetRequestCallback );
+
+    // Find ASIO API index.  TODO: support non-ASIO APIs.
+    PaHostApiIndex apiCount = Pa_GetHostApiCount ();
+    if ( apiCount < 0 )
+    {
+        throw CGenErr ( tr ( "Failed to get API count" ) );
+    }
+    PaHostApiIndex apiIndex = -1;
+    const PaHostApiInfo* apiInfo = NULL;
+    for ( PaHostApiIndex i = 0; i < apiCount; i++ )
+    {
+        apiInfo = Pa_GetHostApiInfo (i);
+        if ( apiInfo->type == paASIO )
+        {
+            apiIndex = i;
+            break;
+        }
+    }
+    if ( apiIndex < 0 || apiInfo->deviceCount == 0)
+    {
+        throw CGenErr ( tr ( "No ASIO devices found" ));
+    }
+    asioIndex = apiIndex;
+
+    lNumDevs = std::min ( apiInfo->deviceCount, MAX_NUMBER_SOUND_CARDS );
+    for ( int i = 0; i < lNumDevs; i++ )
+    {
+        PaDeviceIndex devIndex = Pa_HostApiDeviceIndexToDeviceIndex ( asioIndex, i );
+        const PaDeviceInfo* devInfo = Pa_GetDeviceInfo ( devIndex );
+        strDriverNames[i] = devInfo->name;
+    }
+}
+
+int CSound::Init ( const int iNewPrefMonoBufferSize )
+{
+    if ( deviceIndex >= 0 )
+    {
+        long minBufferSize, maxBufferSize, prefBufferSize, granularity;
+        PaAsio_GetAvailableBufferSizes ( deviceIndex, &minBufferSize, &maxBufferSize, &prefBufferSize, &granularity );
+        if ( granularity == 0 ) // no options, just take the preferred one.
+        {
+            iPrefMonoBufferSize = prefBufferSize;
+        }
+        else if ( iNewPrefMonoBufferSize > maxBufferSize )
+        {
+            iPrefMonoBufferSize = maxBufferSize;
+        }
+        else if ( iNewPrefMonoBufferSize < minBufferSize )
+        {
+            iPrefMonoBufferSize = minBufferSize;
+        }
+        else
+        {
+            // Requested size is within the range.
+            int bufferSize = minBufferSize;
+            while ( bufferSize < iNewPrefMonoBufferSize )
+            {
+                if ( granularity == -1 ) // available buffer size values are powers of two.
+                {
+                    bufferSize *= 2;
+                }
+                else
+                {
+                    bufferSize += granularity;
+                }
+            }
+            iPrefMonoBufferSize = bufferSize;
+        }
+    }
+    else
+    {
+        iPrefMonoBufferSize = iNewPrefMonoBufferSize;
+    }
+
+    vecsAudioData.Init ( iPrefMonoBufferSize * 2 );
+    if ( deviceStream && deviceIndex >= 0 )
+    {
+        LoadAndInitializeDriver ( deviceIndex );
```

Comment:
Calling this function is not supposed to be done in the Init function. LoadAndInitializeDriver should only be called if a new sound card device is selected. The Init function is called very often on any configuration change of the current sound card (like mono/stereo, buffer sizes, etc.).

Reply:
I guess I could rename `LoadAndInitializeDriver(int)` (as `ReloadCurrentDriver` perhaps?) to make things clearer, but since the portaudio stream needs to be re-opened to apply a new buffer size, that call is necessary.